File: 0005. Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        if(n==0 || n==1) return s;
        vector<vector<int>> b(n,vector<int>(n,0));
        int start = 0;
        int len = 1;
        for(int i = 0; i < n; i++){
            b[i][i] = 1;
        }
        for(int i=0; i< n-1; i++){
            if(s[i]==s[i+1]){
                b[i][i+1] = 1;
                start = i;
                len = 2;
            }
        }     
        for(int l = 3; l <= n; l++){
                int j= l-1;
            for(int i = 0; i<n-j; i++){
                
                if(s[i] == s[i+j] && b[i+1][i+j-1]){
                    
                 b[i][i+j] = 1;
                 start = i;
                 len = l;   
                }
            }
        }
        
        return s.substr(start,len);
        
        
    }
};

File: 0005. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""
        dp = [[0]*len(s) for _ in range(len(s))]
        for i in range(len(s)):
            dp[i][i] = 1
        start = 0
        end = 0
        for i in range(len(s)-1, -1, -1):
            for j in range(i+1, len(s)):
                if s[i] == s[j] and (i+1 == j or dp[i+1][j-1] == 1):
                    dp[i][j] = 1
                    if end-start+1 < j-i+1:
                        start=i
                        end=j
        return s[start:end+1]
